halo2_proofs/src/plonk/circuit.rs

```diff
@@ -668,6 +777,60 @@ impl<F: Field> Expression<F> {
     }
 }
 
+impl<F: std::fmt::Debug> std::fmt::Debug for Expression<F> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Expression::Constant(scalar) => f.debug_tuple("Constant").field(scalar).finish(),
+            Expression::Selector(_) => panic!("virtual selectors are removed during optimization"),
```

Comment:
This panic is incorrect; there is no guarantee we won't use the `Debug` impl on `Expression` before optimization. This needs to be implemented.